### DISCUSSION TITLE
help using the adsrgain

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [pschatzmann/arduino-audio-tools#274] [Q&A] [answered]
So, i've looked at examples and I have a sketch that does mostly what I want.

I can trigger a note, set the frequency but there is not noticable envelope now matter which settings I use. I've got an AudioEffects stream. I set the frequency accordingly for the sineWave. I have tried many setting parameters and I can't get a slow attack or decay or release - its always just the same as without ADSR. I have built quite a few arduino synth engine projects so I know what dynamics would typically sound like.
The sustain parameter does set the volume of the sound as one would expect ...
```

SineWaveGenerator<int16_t>    sineWave(32000);  // previously sineWave(32000);    // subclass of SoundGenerator with max amplitude of 32000
AudioEffects<SineWaveGenerator<int16_t>>    effects(sineWave);                             // added for step 2 testing
GeneratedSoundStream<int16_t> sound(effects);                   // originally GeneratedSoundStream<int16_t> sound(sineWave); 
I2SStream out; 
StreamCopy copier(out, sound);   
```
In the setup i set it's ADSR parameters even though they are set on creation of the effect

```
  effects.addEffect(adsr); 
  attackP = 0.7f;
  decayP = 0.1f;
  sustainP = 0.2f;
  releaseP = 0.01f;
  adsr.setAttackRate(attackP);
  adsr.setDecayRate(decayP);
  adsr.setSustainLevel(sustainP);
  adsr.setReleaseRate(releaseP);
```

Further on in the code I stop and start the note playing
```
 adsr.keyOn();
adsr.keyOff();
```
From this definition I figure I need to read and write from the effect because it doesn't seem to be happening automatically.
https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_a_d_s_r_gain.html

```
effect_t process (effect_t input)
  | calculates the effect output from the input

```

The closest thing I found in the documentation so far is something that seems like it. I've looked at many examples and I can't find anything calling process. 
```
uint8_t buffer[1024];
void loop() {
   size_t read = in.readBytes(1024);
   out.write(buffer, read);
}
```
Yes the copier function is explained to do this stream in to out writing but simplifies things by making sure unwritten bytes are written.

I'm going to come back and document how I go in case it's useful. I tried to RTFM so far and got far but this one has me a bit uncertain what to do.

**Comment by pschatzmann** [comment]
Did you already read the [related documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects)?
I suggest that you have also a look at 
- streams-synthbasic1-audiokit
- streams-synthbasic2-audiokit
- streams-synthbasic3-audiokit

in https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit

**Reply by MichaelPNolan** [reply]
I read those 3 sets of example based on audiokit - which I am not using.
I didn't read well enough the "related documentation" since I read other documentation which seems to be "related" but is not the same. Which I explained - so perhaps you could just say "this might help" rather than imply that I don't read the documentation because I've been looking through it as exhaustively as I could for days before I ask for some help.

I've just re-read it and nothing that I haven't done - but as I said I'm not building a synth using your audiokit. I've managed to make one that works without 

```
#include "AudioLibs/AudioKit.h"
AudioKitStream kit;
```
given that in your documentation 

```
#include "AudioTools.h"

SineWaveGenerator<int16_t> sine;
AudioEffects<SineWaveGenerator<int16_t>> effects(sine);
GeneratedSoundStream<int16_t> in(effects); 
I2SStream out;
StreamCopy copier(out, in); 
```
I worked with that - but I guess what you are saying is that without this AudioKit the effects aren't really going to work properly without saying it directly. However, I'd like to point out that your synth example doesn't have an I2SStream out;
Instead it has this very vague ... AudioKit which I guess encapusulates that but in all my other audio projects I like to have something clear that is i2s where I configure it's pins. I'm not using the AIThinker ESP32 Audiokit (which I do own and used for various other synth projects - other libraries or my own work). 

In this case I'm using an ESP32-wroom-32e (which has bluetooth) instead of the the AIThinker audio kit board) i've got a discrete i2s board. Maybe I'm barking up the wrong tree trying to adapt this to something that isn't "example hardware". So trying to use AudioKitStream seems to be ... ill advised, but sure I tried it and got a lot of errors with wire and "please verify your AUDIOKIT_BOARD setting" error is not exactly covered in your related documentation which doesn't even really seem to mention that it's got that target hardware.
I'm using i2c for a couple of devices and I'm trying to integrate a few ideas. I feel like I've used your code correctly.

![esp32 wroom 32e-small](https://user-images.githubusercontent.com/48641859/180948779-78bf25f7-3788-4379-b5c2-f12c110bd881.jpg)

You know, in coding people make bugs and they often don't see them until a friend reads it or they themselves ask for help and I am simply trying to go through that egoless debugging. What I can say is **that I'm actually asking for an example of the adsr.process function which seems highly elusive**. Can you tell me it's not something that the end user of these libraries needs to use so I can drop it or else maybe help me out? 

You have helped and I massively appreciate it. Yes, we are all idiots compared to the more developed authors like you when we start the journey of familiarizing ourselves with so many different examples. Is it so much to ask that maybe doing it differently just requires some info and that maybe with i2sstream this effect thing isn't working completely as it the AudioKit object does.

**Comment by pschatzmann** [comment]
I am using the [AudioKit](https://github.com/pschatzmann/arduino-audiokit) for many examples because it is just too much work to wire up I2S every time I do a new sketch.

The important thing you can get from all examples is how things are supposed to work together.

If you don't have an AudioKit you can replace the AudioKitStream in any example with any supported output class: in your case I2SStream and it should work as well. Please note that the audio in the examples gets triggered by the keys on the AudioKit. I don't know how you planned to implement the triggering of the sound.

As you should be able to see from the example, there is no need to mess with the process function but you should use keyOn and keyOff which could easily be triggered from midi events. Please note that timing is important here: if you want the sound to be for 1 second you need to wait 1 second after calling keyOn before calling keyOff.

You can copy the audio data yourself as you did in the loop, or better just call copier.copy(), since you already defined the StreamCopy copier in your sketch.

**Reply by MichaelPNolan** [reply]
I really works - but there is no envelope effect as defined in the effect.
I'm not saying it doesn't work - there is sound ... but no envelope.
I do use KeyOn and KeyOff - that works, so yes, that means the effect is functional .. but not, fully functional.
Yes I do already have a timing between KeyOn and KeyOff ... and I have other projects that use Midi (serial and USB) and I even wrote my own midi libraries from the standard. So what might help me is to know how does it actually modify the dynamics of the sinewave because ... other than on and off it is has no sloped dynamics.

```
void loop() {
   copier.copy();
  updateCounter();
  
  if (lastCheck != hundredths){ //only check on a new hundredths of sec
    checkButtons();
    if (hundredths == 10 || noteTestFlag){ //its a looping cycle of 0-99 so any number will happen once a secon
      if(playNoteOn){
        playNoteOn = false;
        Serial.println(" off");
        //sineWave.end();
        adsr.keyOff();
        //sound.end();
      } else {
        if(noteTestFlag){
          playNoteOn = true;
          noteTestFlag = LOW;
          sineWave.setFrequency(noteFrequency);
          //sound.begin(); //this is a stream with effects that has the sineWave object
          adsr.keyOn(0.5);
        #ifdef LCD_I2C
          lcd.clear();
          lcd.setCursor(0,1); //col the row
          lcd.print("note#: ");  lcd.print(String(64+noteModifier));
        #else
          Serial.println(String(64+noteModifier++));
        #endif
        } else {
          checkCurrent();
        }
      }
    }
    lastCheck = hundredths;
  }
    
}
```

**Comment by pschatzmann** [comment]
Did you try with the ADSRGain default values which are
        ADSRGain(float attack=0.001, float decay=0.001, float sustainLevel=0.5, float release= 0.005, float boostFactor=1.0)
With this the effect should be better audible...

If you check the ADSRGain class the update is delegated for each tick (=sample) to the ADSR class.

**Reply by MichaelPNolan** [reply]
Yes, 
         I did try that - one of the examples didn't have the boostFactor variable and that seems to work (4 parameters) but later I added Boost Factor. I used it with default values as shown. 
I can change sustain and, as with my other synths the volume of the note changes but however I change attack, decay or release there is no ramping up or down or release play time change.
In case it helps - I'm 53, I have a degree in computer science, i've worked in many fields of computing. I'm not an amazing programmer or anything but often find bugs and report them on Git projects and have my suggested changes added to the code base. I'm trying always to be useful and I'm sorry that you have to work through to get a sense of where I am.
If you want me to go off and build this with another hardware to see if the ADSR works - I suspect it will. However, if there is something that has to happen so that the effect actually calculates and applies the dynamics over time it would be good to know that.

I can't help but want to know what calls? So that it can change the dynamics?

```
        effect_t process(effect_t input) {
            effect_t result = factor * adsr->tick() * input;
            return result;
        }
```

**Comment by pschatzmann** [comment]
The dynamics is implemented in the ADSR class and executed via adsr->tick() 
The process method is called by AudioEffects class to process the chain of potentially multiple effects

I was just double checking the synthbasic2 example and I can hear the ramp up, so I am not sure what your issue could be: Maybe you can also add a  Serial.println(" on"); to double check that the on function is called only once...
Did you remove the redundant setter for the adsr values in the setup ?

**Reply by MichaelPNolan** [reply]
Yes, I commented out the commands where I was confirming the settings of ADSL. 
I added that debug code you suggested to see if "on" is triggering more than 1 time.

So when I press a button it increments or decrements (depending which button) the note modifier. Elsewhere in the code it calls "noteFrequency = notes.frequency(64+noteModifier++);" for example to increment. Your notes object gives the frequency from the scale.
Its only test code but there is usually some interval of under 1 second before it calls Note off ... so the note has time to be heard fading in or slowly releasing but its a very gated on/off sound. 

```
Button 1 pressed
 on
 off
Button 0 pressed
 on
 off
Button 0 pressed
 on
 off
Button 0 pressed
 on
 off
Button 1 pressed
 on
 off

```

This is the setup code and yes there are a few devices attached but ... well I don't see that their basic comms would interfere with your libraries. All I can think is that if it seems right that I should try with the AIThinker hardware and adapt this code to that to see if there is a different result if I stop using i2sStream and replace with AudioKitStream type  ... because, well, I suspect something is different with these but I have no evidence.

```
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  //while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  setupCurrentSensor();
 //setupNeopixel();
 //pixelRed();
#ifdef LCD_I2C
  lcd.init();
  lcd.backlight();
  lcd.clear();
#endif
  // setup effects  - step 2 testing
  effects.addEffect(adsr); 
  adsr.setActive(HIGH);
  attackP = 0.9f;
  decayP = 0.1f;
  sustainP = 0.2f;
  releaseP = 0.99f;
  //adsr.setAttackRate(attackP);
  //adsr.setDecayRate(decayP);
  //adsr.setSustainLevel(sustainP);
  //adsr.setReleaseRate(releaseP);
  Serial.println(String(adsr.attackRate()));
  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.pin_bck = PIN_I2S_BCK; //14;
  config.pin_ws = PIN_I2S_WS; //15;
  config.pin_data = PIN_I2S_DATA_OUT;// 27;
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.bits_per_sample = 16;
  sineWave.begin(config, 0);
  out.begin(config);
  sound.begin(config);                                         
  
  Serial.println("started...");
 

  buttonSetup();
}
```

**Reply by MichaelPNolan** [reply]
I have loaded "streams-synthbasic2-audiokit"
and used AUDIOKIT_BOARD 6 (the AC101 version) and yes - it works the sound is a ramping up tone. That's with the code "as is" so I guess I can attempt some kind of way to see if I can make it work with other hardware?

**Comment by pschatzmann** [comment, marked as answer]
I created a simple [streams-adsr-i2s.ino](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-adsr-i2s/streams-adsr-i2s.ino) example. 
Can you check if this is working for you with I2S ?
I only tested this on the AudioKit....

**Reply by MichaelPNolan** [reply]
Yes, it works with a ramping up tone ... too and I only made one addition. The pins for my i2s device. (3 lines) 
```
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = PIN_I2S_BCK; //14;
  cfg.pin_ws = PIN_I2S_WS; //15;
  cfg.pin_data = PIN_I2S_DATA_OUT;// 27;
  //i2s.setVolume(1.0);
  i2s.begin(cfg);
```
So, ok this is what I found was different/wrong that made mine not work.
I had.

` auto config = i2s.defaultConfig(TX_MODE);`
Where config was a keyword highlighted in the Arduino editor as a keyword. Which I have been ignoring for some reason.
So i changed it to cfg instead and voila. Many thanks and sorry about that cheif.
It only begs the question - how come that worked to the extent that the i2s interface worked on those pins but not the extent that the config was good for everything? Perhaps because I had this lines in there too.

```
#define PIN_I2S_BCK   14
#define PIN_I2S_WS    15
#define PIN_I2S_DATA_OUT 27
```
Thanks ... so I guess I need quite a few instances of the stream for polyphony.
